interface overhaul

HDembinski opened the pull request.

HDembinski commented:
This is a large patch.

- Closes #31  #5  #29  #27 
- Fixed the doc build
- Added new tests that actually check the bootstrap result and the confidence intervals
- Went back from src/resample layout to ./resample layout
- Added a single location for the version, in ./resample/version.py (mimicing e.g. scipy)
- Parametric bootstrap now also works with 2d data and uses the multivariate normal in this case
- I found a lot of bugs in the parametric bootstrap, which I all fixed (especially poisson was driving me mad)
- I fixed bugs in the confidence interval computation, not sure if I introduced them or whether they were already there
- Jackknife and non-parametric bootstrap work with ND-data
- I started to use pylint in my editor Atom and found it useful. It gives a lot of false positives at first and is very opinionated about style, so one has to tame it with an appropriate pylintrc, which I added. resample/pemutation.py is currently excluded from pylint, as it otherwise would complain a lot about unpythonic variable names. I did not touch permutation.py.
- I adapted docstrings to numpydoc, and also changed the sphinx extension from napoleon to numpydoc (was there a reason to prefer napoleon? I couldn't find a reason)
- I removed `mock` from conf.py, because it was breaking all my tests. I don't know what this was supposed to do.
- I removed one example from the bootstrap.rst, because we need to have all the examples in the docstrings, so that they are visible to those who use `help()` and tested by doctest.
- I had to increase scipy minimum version to 1.5, because I found a bug in scipy.stats.poisson, that was only fixed in version 1.5.
- I renamed bootstrap_ci to confidence_interval. The name is negotiable, but it is better to have somewhat longer self-explanatory names, people can always abbreviate them `from resample.bootstrap import confidence_interval as ci`
- I allowed stratification also for parametric bootstrap. It works also there and it makes sense to use parametric bootstrap with stratification.

I went back from the src/resample layout to the simple resample after reading a bit more about this and also having difficulties with the other layout.
- It seems the blog entry that favoured src/<package> over <package> is outdated and a minority opinion.
- I could not install the package with pip in develop mode in the other layout, for example. Jupyter notebook did not find it.

HDembinski commented:
I forgot: I removed the keyword `family` in `bootstrap` and merged it with `method`.

HDembinski commented:
More things I forgot: `confidence_interval` (formerly bootstrap_ci) now sets the `ci_method` as before, but the bootstrap method via `method`. This is less clear, but allows for greater automatic consistency in our interface: functions that use `bootstrap.resample` can just forward extra keyword arguments which they do not consume themselves to resample. This way, we don't need to repeat the arguments for the resampler all the time. Redundancy is bad, [we aim to be DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

dsaxton commented:
Thanks! I will take a look more later on this week. (I'm assuming it would be too much work at this point to break this up into pieces that are more or less orthogonal to one another?)

> * I found a lot of bugs in the parametric bootstrap, which I all fixed (especially poisson was driving me mad)
> * I fixed bugs in the confidence interval computation, not sure if I introduced them or whether they were already there

Can you describe what the bugs were and did we add tests for those cases?

> * I adapted docstrings to numpydoc, and also changed the sphinx extension from napoleon to numpydoc (was there a reason to prefer napoleon? I couldn't find a reason)

No reason. When I added this I was (and still am) not very familiar with how Sphinx works, so those were likely just defaults.

> * I renamed bootstrap_ci to confidence_interval. The name is negotiable, but it is better to have somewhat longer self-explanatory names, people can always abbreviate them `from resample.bootstrap import confidence_interval as ci`

Agreed.

> * I allowed stratification also for parametric bootstrap. It works also there and it makes sense to use parametric bootstrap with stratification.

How is this working? I'm guessing we fit separate distributions for each stratum and sample from those?

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+        Original sample.
+    size : int, optional
+        Number of bootstrap samples to generate. Default is 1000.
+    method : str or None, optional
```
Good idea to move the parametric bootstrap into this argument

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+        "student": stats.t,
+    }.get(method, None)
+
+    if dist is None:
```
I wouldn't be opposed to being more strict and always raising here rather than potentially (albeit probably not very likely) accidentally using the wrong thing from scipy.stats. The docstring is already explicit about the allowable  distributions.

HDembinski replied:
The idea was that one can use any distribution from scipy.stats, even though not all are advertised. We only advertise those that are tested. In principle, I see no reason to restrict the supported parametric families.

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+        try:
+            dist = getattr(stats, method.lower())
+        except AttributeError:
+            raise ValueError("Invalid family: '{}'".format(method))
```
```suggestion
            raise ValueError(f"Invalid family: '{method}'")
```

HDembinski replied:
I love f-strings, but I thought I cannot use them, because we want to support older Python versions, but it turns out they were already added in 3.6.

HDembinski commented:
> Thanks! I will take a look more later on this week. (I'm assuming it would be too much work at this point to break this up into pieces that are more or less orthogonal to one another?)

I know this is not proper, but we discussed the changes before so I hope it is ok. I had to do this all in one go, there were too many little things to keep track of separately. I rather revert/modify particular changes that you criticize.

> Can you describe what the bugs were and did we add tests for those cases?

Sorry, I didn't keep of every detail, but parametric resampling with poisson did not work at all and the resampled pareto distribution was wrong, I think. I added statistical tests for all the functionality in the bootstrap module. I check that all the resampling methods produce statistical distributions which are compatible with the original data. This goes much beyond what the tests did so far.

> > * I allowed stratification also for parametric bootstrap. It works also there and it makes sense to use parametric bootstrap with stratification.
> 
> How is this working? I'm guessing we fit separate distributions for each stratum and sample from those?

Yes.

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+
+    if sample.ndim > 1:
+        if dist != stats.norm:
+            raise ValueError("family '%s' only supports 1D samples" % method)
```
```suggestion
            raise ValueError(f"family '{method}' only supports 1D samples")
```

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+from resample.empirical import quantile_gen as _quantile_gen
+
+
+def resample(
```
I'm a little confused by this extra layer of abstraction, could we not wrap this logic into the bootstrap function itself? I haven't looked too closely but it seems potentially duplicative given that this is returning bootstrap samples. If we did have this it seems it might make more sense as a private helper function than a user-facing one.

HDembinski replied:
I thought we had discussed this before and agreed to use the `bootstrap` exclusively for "array of function applied to bootstrap replicas". `resample` is not a function but a generator of bootstrap replicas. This split allows one to get one replicated sample at a time instead to do with whatever you want, instead of generating them all at once. We cannot fold this into `bootstrap`, because `bootstrap` is a function. It would be weird to use the sample object as a function and as a generator.

HDembinski replied:
The stratification code looks simpler, now, too.

HDembinski replied:
And it is for the symmetry with jackknife (although that is the least important reason).

Here is another argument for adding the generator separately. When designing an interface, one should ask: what is the minimum amount of useful work that the interface should do for the user. In this case, the minimum amount of useful work is to generate a single replicated sample (either jackknife'd or bootstrapped), which the use can use in any way they want. Getting a single replicated sample at a time is better than an interface that precomputes all samples and only allows me to get them all at once. Because in the latter design, I have to hold all the replicas in memory at once, while in the former design I only need to hold on replica in memory.

In case of the parametric and ordinary bootstrap, there are large savings in memory consumption when you use a generator. Python loves generators for that reason, they replaced a lot of the former sequences with generators (e.g. `range`). They are more memory efficient than fully realized sequences.

HDembinski replied:
Ok, but I think the generator should be user-facing. We are exposing something useful. For example, if I want to apply my function in parallel on all the resampled samples, I would do something like this:
```py
from concurrent.futures import ProcessPoolExecutor as Pool
def myfn(x): ...
data = ...
with Pool() as p:
    for  replicas in p.map(myfn, resample(data)):
        # do stuff with replicas
```

HDembinski replied:
The resample generator is really the core element of this library, everything else is sugar and convenience on top.

HDembinski replied:
Actually..., I think I need to fix the jackknife generator so that returns copies by default, otherwise the above code would not work for the jackknife. The inplace modification should be opt-in to increase performance when possible.

dsaxton replied:
Yeah, below I convinced myself that it makes sense to expose it (and remembered we'd mentioned it before)

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+import numpy as np
+from scipy import stats
+from resample.jackknife import jackknife as _jackknife
+from resample.empirical import quantile_gen as _quantile_gen
```
I'd personally rather use something like `quantile_function` or `quantile_fn` since the `_gen` sounds a bit peculiar (it makes it sound like it's a generator but that's not what we mean). Similar to `confidence_interval` a user can always abbreviate `quantile_function` to something shorter.

HDembinski replied:
I chose `..._gen` because Scipy does uses it. They have `poisson_gen` to generate a `poisson` object, for instance (in stats/_discrete_distns.py). But if you prefer `_fn` then I will change this.

HDembinski replied:
Changed to _gen.

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+    sample : array-like
+        Original sample.
+    **kwds
+        Keywords are forwarded to `resample`.
```
Again I feel this might be unnecessary cognitive overhead for a user, who now has to think in terms of two functions / signatures instead of the one they're using

HDembinski replied:
I hope I can convince you otherwise.

HDembinski replied:
As explained above, I think that the resample generator is the core of the library, not the `bootstrap` or the `jackknife` functions. Instead of forwarding the arguments to the resampler, we could also pass the generator instance to those methods. With the generator factorized out, the functions `jackknife` and `bootstrap` essentially do the same thing.

dsaxton replied:
Is there a benefit to passing them as kwargs instead of explicitly, aside from coding convenience? Personal preference but I find kwargs a bit opaque in general and usually would rather clearly enumerate what a function needs in order to work.

HDembinski replied:
The benefit is that there is no redundancy. Redundancy in code is bad:
https://de.wikipedia.org/wiki/Don%E2%80%99t_repeat_yourself.

This is not about coding convenience, I design interfaces with the user in mind, but there are always trade-offs. If there is no redundancy, there is no possibility of information going out of sync. There is no possibility for keywords to differ in `resample` and the other methods that use it. The burden for the user to look up another function is small, and offset by the fact that the total documentation gets more compact, because the same information is not repeated.

There is another design, which avoids this problem and is even more elegant:
`def bootstrap(fn, resampler): ...`

Instead of passing the arguments for the `resample` generator, we pass an instance of the `resample` generator. There is no redundancy then in the docs and the interface is clean and simple.

dsaxton replied:
We can leave it this way but I'd rather use the convention of calling them `kwargs` instead of `kwds` (can be a follow-up)

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+    if method == "ordinary":
+        return _resample_ordinary(sample, size, rng)
+
+    dist = {
+        # put aliases here
+        "gaussian": stats.norm,
+        "normal": stats.norm,
+        "log-normal": stats.lognorm,
+        "log-gaussian": stats.lognorm,
+        "inverse-gaussian": stats.invgauss,
+        "student": stats.t,
+    }.get(method, None)
```
This is much better than how this was done before 👍

HDembinski replied:
I hope it is ok that I added a few alias for families. For the normal distribution, "norm", "normal", and "gaussian" are accepted, for example. For student's t, "student" and "t" are accepted, etc.

HDembinski commented:
Regarding the bugs. I am looking into the previous code now and I see that you had already a workaround for one of the poisson issue. I was not aware that `poisson` does not have a `fit` method, and when I tried to write the fitting code in an universal way, I stumbled over this. So poisson was fine before, I introduced the bug when I refactored the code and then later fixed it.

Regarding pareto, the issue there was two-fold. Firstly, pareto in scipy-1.4 did not support passing numpy.random.Generator as `random_state`, while all others do support that. I only discovered this when I tried to use numpy.random.Generator, so the previous code which only accepted an `int` was also fine. Then I had trouble with differences in the random numbers produced that the scipy pareto object produced in the unit test for the parametric bootstrap. In either case, the code before was fine, I think. The bugs appeared during my refactoring.

dsaxton commented on setup.cfg:
```diff
@@ -1,8 +1,18 @@
+exclude = setup.py
+
+[pylint]
+rcfile = .pylintrc
```
I'm actually not familiar with pylint. What is this adding on top of black / flake8?

HDembinski replied:
pylint also complains about stylistic issues, one-letter variables, for example.
https://www.reddit.com/r/Python/comments/82hgzm/any_advantages_of_flake8_over_pylint/

HDembinski replied:
I switched my linter in Atom to flake8 now and I like it better than pylint. May I remove the pylint stuff? I also like the philosophy of flake8 better.

HDembinski replied:
Removed pylint

dsaxton commented on resample/version.py:
```diff
@@ -0,0 +1 @@
+version = "0.21"
```
Should we just bump this to 1.0.0? The current version number is essentially meaningless.

HDembinski replied:
Yes, once we finished this interface overhaul, I am all for changing to 1.0.0!

dsaxton commented on resample/empirical.py:
```diff
@@ -0,0 +1,80 @@
+from resample.jackknife import jackknife as _jackknife
+
+
+def cdf_gen(sample: np.ndarray) -> Callable:
```
I think this could just be `cdf`?

HDembinski replied:
It returns a function, you said before you wanted those to be called "X_fn". If that is called just `cdf`, then also `quantile_fn` should just be `quantile`. But I think you were right that higher-order functions which return other functions should have a different name than ordinary functions.

dsaxton replied:
It stands for cumulative distribution function so "function" is already in the name

HDembinski replied:
Yes, but `cdf_gen` creates a cdf, it does not return the value computed by the cdf. It is a cdf generator / higher-order function / meta-function, whatever you prefer to call it. Likewise for `quantile`.

The usual naming conventions in programming is that we do not call something a "function" or a "mapping" explicitly in the function name. Therefore, something that computes the cumulative density would normally be called `cumulative_density` or `cd`, not `cumulative_density_function` or `cdf`. However, since `cdf` is an abbreviation that everyone recognizes, libraries usually chose `cdf`, making the "function" part of the name, although it goes against normal conventions.

This is the situation we are facing here. Consistency is important. In scipy, norm.cdf(...) returns the cumulative density for the argument, not the cdf. norm.cdf is a normal function. Our `cdf_gen` is a metafunction. We want to make that clear in the name. It is therefore not superfluous to add `_gen` or `_fn` to indicate that this is a function generator.

Having said all that, I think the suffix `cdf_gen` makes that distinction more clear than `cdf_fn`.

HDembinski replied:
In C++, could be considered a factory `cdf_gen`, and then an appropriate name would be `make_cdf`, but I don't think this is a good match for Python.

dsaxton replied:
I was actually thinking in terms of the name of the thing outside the world of programming and not the return type when suggesting we use a name like quantile_function (communicating that it's a higher-order function in my mind is the purpose of documentation / type annotations and generally wouldn't be the responsibility of the function name, similar to how we wouldn't name something quantile_float if it was returning a single quantile).

My personal opinion and maybe we just merge for now and reconsider later.

HDembinski replied:
Ok, let's discuss this later.

dsaxton commented on resample/jackknife.py:
```diff
@@ -1,31 +1,50 @@
-    yield x.view(x.dtype)  # must return view to avoid a numba life-time bug
+def resample(sample: Sequence) -> np.ndarray:
+    """
+    Generator of jackknife'd samples.
```
```suggestion
    Generator of jackknifed samples.
```

dsaxton commented on resample/jackknife.py:
```diff
@@ -1,31 +1,50 @@
-        x[i] = a[1 + i]
-    # yield x0
-    yield x.view(x.dtype)  # must return view to avoid a numba life-time bug
+def resample(sample: Sequence) -> np.ndarray:
```
Thinking about this some more I think it's probably okay / good to add to these to the API as I could see someone wanting to use them on their own, so I'd say it's more about making the hand-off of data to these functions less visible when they aren't being used directly.

dsaxton replied:
Actually, I do see that this was mentioned in an issue last month, must have slipped my mind

dsaxton commented on resample/jackknife.py:
```diff
@@ -1,31 +1,50 @@
+    ----------
+    sample: array-like
+        Sample. If the sequence is multi-dimensional, the first dimension must
+        walk over i.i.d. observations.
```
Being a bit pedantic, but would it be more correct to simply call these "the individual observations" or something like that rather than "i.i.d. observations" since we don't in general know the statistical properties of the data (maybe they're in fact dependent)?

HDembinski replied:
If the bootstrap/jackknife are to work correctly, the observations must be i.i.d., right? (ignoring stratification for a moment, which in the end is just a way to fix that stratified data is not i.i.d., but consists of classes of i.i.d. observations)

dsaxton commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+            yield dist.rvs(size=n, random_state=rng)
+
+
+def _confidence_interval_percentile(
```
Thanks for breaking this into helper functions, I think it improves readability / maintainability a lot

HDembinski replied:
Thanks!

I sometimes also write long functions, instead of factoring things out, like one should. You asked before what pylint does on top of flake8. pylint, for example, complained here about the original function being too long and having too many 'if's, and thus encouraged me to factor things out into smaller pieces.

HDembinski commented:
Regarding pylint: I am happ to remove it again. While working on the refactoring, it was useful to have, but perhaps flake8 as a linter would have worked for me as well. I just happened to chose pylint.

dsaxton commented on resample/tests/test_bootstrap.py:
```diff
@@ -0,0 +1,224 @@
+
+
+@pytest.mark.parametrize("method", NON_PARAMETRIC | PARAMETRIC_CONTINUOUS)
+def test_resample_1d_parametric(method, rng):
```
Is there any way to simplify this test? Also we may want to rename since it's not only for parametric sampling.

HDembinski replied:
I always aim to write the most compact code to achieve some goal, so if there is a simpler way, you need to suggest it.

This is a statistical test which checks that the resampling actually produces distributions which agree with the original distribution. To compare the realized with the expected distributions, I use a chi-square test. I then check whether the pvalues are flat. I found that this test is very sensitive to any statistical deviations.

I am very satisfied with the result and I spend a lot of time on this. This is the test to convince people that this module works correctly, so it is rather important.

HDembinski replied:
I fixed the name.

dsaxton replied:
I guess it's fine as long as we know it'll fail randomly and that's documented

HDembinski replied:
I made sure it cannot fail randomly by fixing the seed and the RNG. As you can see, the test passes on all machines and all Python versions.

dsaxton commented on notebook/confidence_intervals.ipynb:
```diff
@@ -0,0 +1,121 @@
+{
```
Not sure why this notebook was added to the PR, was this intentional?

HDembinski replied:
Sorry, that was supposed to be added in another PR. I will remove it.

HDembinski commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+    method: str = "balanced",
+    strata: Optional[Sequence] = None,
+    random_state: Optional[Union[np.random.Generator, int]] = None,
+) -> np.ndarray:
```
You are right, I will fix this.
https://stackoverflow.com/questions/38419654/proper-type-annotation-of-python-functions-with-yield

HDembinski replied:
Done

HDembinski commented on resample/bootstrap.py:
```diff
@@ -0,0 +1,285 @@
+        return (np.mean(sample, axis=0), np.cov(sample.T, ddof=1))
+
+    if dist == stats.t:
+        fit_kwd = {"fscale": 1}  # HD: I think this should not be fixed to 1
```
I think it would help to allow `fscale` to vary. Then one can use the parametric bootstrap with student's t also for distributions that are approximately normal, but have somewhat heavier tails. I found that the student's t distribution works well to approximately fit certain resonances. What do you think?

HDembinski replied:
I will remove the comment and rather add an issue to discuss this further, to move on with this PR.

HDembinski commented on resample/empirical.py:
```diff
@@ -0,0 +1,80 @@
+        Empirical influence values.
+    """
+    n = len(sample)
+    return (n - 1) * (fn(sample) - _jackknife(fn, sample))
```
Yes, it does. We should write that requirement down somewhere...

HDembinski replied:
Added a comment to the resampler.

dsaxton commented:
> Regarding pylint: I am happ to remove it again. While working on the refactoring, it was useful to have, but perhaps flake8 as a linter would have worked for me as well. I just happened to chose pylint.
> 
> Edit: We could also keep the .pylintrc checked in, but not use pylint in Ci.
> 
> Edit: I removed pylint.

Yeah I think removing for now is good and could be revisited later. If there are some extra coding conventions that it enforces that we consider good (e.g., not using single letter variable names) we could always just enforce ourselves instead of trying to incorporate into the CI and potentially bringing on other baggage from the tool (I'm personally not familiar with it yet so I don't know what that might be).

HDembinski commented:
> Yeah I think removing for now is good and could be revisited later. If there are some extra coding conventions that it enforces that we consider good (e.g., not using single letter variable names) we could always just enforce ourselves instead of trying to incorporate into the CI and potentially bringing on other baggage from the tool (I'm personally not familiar with it yet so I don't know what that might be).

I think one can also add additional checks to flake8. I am new to these tools, so I haven't fully explored all the options there. pylint is pretty slow compared to flake8. That does not matter for CI, but it matters when you turn it on in your editor for linting. I am now in the camp flake8. :)

HDembinski commented:
It looks like we can wrap this up soon, in time for the PyHEP presentation. The workshop starts on July 13th.

dsaxton commented on resample/__init__.py:
```diff
@@ -0,0 +1,5 @@
+resample package.
+"""
+
+__version__ = "0.21"
```
```suggestion
__version__ = "1.0.0"
```

Just update this now?

HDembinski replied:
I usually increase the version number when I make a new release. If I am happy with the code state of the repository for a release, I make a special release PR in which I only increase the version number and write the changelog. For me this has dedicated advantages, which do not matter here. Can we perhaps jump to 0.9 instead, until we are completely finished with the interface update? There are two more things to settle
- the naming of the functions in `empirical.py`
- how to support precomputed pre-computed bootstrap and jackknife estimates

HDembinski replied:
Actually, that this version string in `__init__.py` is still there was a mistake, I fixed that now. I also get why we cannot switch to 0.9, the last release you made on PyPI was 0.21, which seems to be a typo, but it will rank higher than 0.9, I presume...

HDembinski replied:
Implemented.

dsaxton commented on resample/jackknife.py:
```diff
@@ -0,0 +1,185 @@
+    the generator state, by setting the `copy=False`. However, this will only produce
+    correct results if the generator is called strictly sequentially in a single-
+    threaded program and the loop body consumes the view and does not try to store it.
+    The following program shows that happens otherwise:
```
```suggestion
    The following program shows what happens otherwise:
```

dsaxton commented on resample/jackknife.py:
```diff
@@ -0,0 +1,185 @@
+        Sample. If the sequence is multi-dimensional, the first dimension must
+        walk over i.i.d. observations.
+    copy: bool, optional
+        If true, return the replicated sample as a copy, otherwise return a view into
```
```suggestion
        If `True`, return the replicated sample as a copy, otherwise return a view into
```
Also consistent with `False` later in the docstring

dsaxton commented on resample/version.py:
```diff
@@ -0,0 +1 @@
+version = "0.21"
```
```suggestion
version = "1.0.0"
```

dsaxton commented on .github/workflows/ci.yml:
```diff
@@ -20,10 +20,11 @@ jobs:
-        pip install black flake8 mypy
+        python -m pip install -r requirements.txt
+        python -m pip install black flake8 mypy
     - name: Format
```
```suggestion
    - name: black
```
Might as well do this for other checks and not just flake8

dsaxton commented on .github/workflows/ci.yml:
```diff
@@ -20,10 +20,11 @@ jobs:
+      run: black --check ./resample
+    - name: flake8
+      run: flake8 ./resample
     - name: Type check
```
```suggestion
    - name: mypy
```

HDembinski commented:
@dsaxton Please merge if you think it is ready.
